### PR TITLE
sane: Add support for the unfree Fujitsu ScanSnap drivers

### PR DESCRIPTION
**Pull request by Profpatsch**
This adds the scanner files already linked from the
`etc/sane.d/epjitsu.conf` file, which are extracted from the Windows
drivers and mirrored on GitHub.

Being a Japanese hardware vendor, Fujitsu’s software release &
licensing methods are horrifying, but their scanners are some of the
best, so we should definitly have discoverable support for them, which
this patch hopefully adds.

Inspiration was taken from the following sources:
https://www.josharcher.uk/code/install-scansnap-s1300-drivers-linux/
https://ubuntuforums.org/archive/index.php/t-1461915.html
https://github.com/stevleibelt/scansnap-firmware

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/hardware/sane.nix [hunk 1]
```diff
@@ -98,6 +102,34 @@ in
       '';
     };
 
+    hardware.sane.drivers.scanSnap = mkOption {
+      type = types.submodule ({ config, ... }: {
```
Comment (sternenseemann) [inline nixos/modules/services/hardware/sane.nix:106]: Why are you using a submodule here?

File: nixos/modules/services/hardware/sane.nix [hunk 2]
```diff
@@ -4,7 +4,11 @@ with lib;
 
 let
 
-  pkg = pkgs.sane-backends;
+  pkg = pkgs.sane-backends.override
+    (lib.optionalAttrs config.hardware.sane.drivers.scanSnap.enable {
+      scanSnapDriversUnfree = true;
```
Comment (sternenseemann) [inline nixos/modules/services/hardware/sane.nix:9]:
```suggestion
      scanSnapDriversUnfree = config.hardware.sane.drivers.scanSnap.enable;
```

File: nixos/modules/services/hardware/sane.nix (hunk 1 above)
Comment (Profpatsch) [reply nixos/modules/services/hardware/sane.nix:106]: idk, brainfart. The module system is too complicated for my mouse brain.

File: nixos/modules/services/hardware/sane.nix (hunk 2 above)
Comment (Profpatsch) [reply nixos/modules/services/hardware/sane.nix:9]: Hrm I can just leave out the optionalAttrs now that I think about it. Good catch.

Comment (sternenseemann) [reply nixos/modules/services/hardware/sane.nix:9]: Oh, I missed those. I guess you could also leave it as is.

Not sure if we can have a tiny performance gain by only overriding if unfree, but it's the module system, so I'd say nicer code is more important.

**Review by sternenseemann — APPROVED** [review]

**Comment by symphorien** [comment]
Other sane drivers are packaged and built separately from sane (see epkowa for example) and then you are supposed to add them to sane with hardware.sane.extraBackends. Would it be possible to do this instead or is scanSnap too weird?

**Comment by Profpatsch** [comment]
> Other sane drivers are packaged and built separately from sane (see epkowa for example) and then you are supposed to add them to sane with hardware.sane.extraBackends.

That’s not the same as far as I can understand. Look at the code, it’s put in a different location and a different file forma.

**Comment by symphorien** [comment]
Hmm indeed this is just providing firmware to a backend that is already distributed. Thanks for the clarification.

**Comment by Profpatsch** [comment]
Gonna merge now, can confirm it’s working on my machine.

heads up, libreoffice depends on `sane-backends` for some ungodly reason, so this change will make it rebuild.

**Pull request merged**